x-pack/plugins/security_solution/public/resolver/store/data/selectors.ts

```diff
@@ -167,6 +168,105 @@ export function hasMoreAncestors(state: DataState): boolean {
   return tree ? resolverTreeModel.hasMoreAncestors(tree) : false;
 }
 
+/**
+ * A map of `entity_id`s to functions that provide information about
+ * related events by ECS `.category` Primarily to avoid having business logic
+ * in UI components.
+ */
+export const relatedEventInfoByEntityId = createSelector(
+  relatedEventsByEntityId,
+  relatedEventsStats,
+  function selectLineageLimitInfo(
+    /* eslint-disable no-shadow */
+    relatedEventsByEntityId,
+    relatedEventsStats
+    /* eslint-enable no-shadow */
+  ) {
+    const relatedEventInfoEntries = [...(relatedEventsStats || new Map()).entries()].map(
+      ([entityId, stats]) => {
+        const eventsResponseForThisEntry = relatedEventsByEntityId.get(entityId);
+        const hasMoreEvents =
+          eventsResponseForThisEntry && eventsResponseForThisEntry.nextEvent !== null;
+        /**
+         * Get the "aggregate" total for the event category (i.e. _all_ events that would qualify as being "in category")
+         * For a set like `[DNS,File][File,DNS][Registry]` The first and second events would contribute to the aggregate total for DNS being 2.
+         * This is currently aligned with how the backed provides this information.
+         *
+         * @param eventCategory {string} The ECS category like 'file','dns',etc.
+         */
+        const getAggregateTotalForCategory = (eventCategory: string): number => {
+          return stats.events.byCategory[eventCategory] || 0;
+        };
+
+        /**
+         * Get all the related events in the category provided.
+         *
+         * @param eventCategory {string} The ECS category like 'file','dns',etc.
+         */
+        const getMatchingEventsForCategory = (eventCategory: string): ResolverEvent[] => {
+          if (!eventsResponseForThisEntry) {
+            return [];
+          }
+          return eventsResponseForThisEntry.events.filter((resolverEvent) => {
+            for (const category of [allEventCategories(resolverEvent)].flat()) {
+              if (category === eventCategory) {
```

Comment:
this runs for all the categories on all the related events for a node right? do you think it would be worth wrapping `getMatchingEventsForCategory` in something like [`defaultMemoize`](https://github.com/reduxjs/reselect#defaultmemoizefunc-equalitycheck--defaultequalitycheck)

The function will be memoized for the last category. Since the function is allocated on a per-node level, the memoize will be per-node. It keeps the result of the last passed arg.

Reply:
Like this? https://github.com/elastic/kibana/pull/70097/commits/640f721ac2691f5157769cb1770136d9d3f1f7cd